Build Failed (with gradle plugin)

build failed [(with fork plugin)](https://github.com/shazam/fork-gradle-plugin) using the command:

`./gradlew fork`

Also the test report output was useless (see screenies) I got this stacktrace from the command line.
(geny motion emu)

Build passes with:

`./gradlew connectedAndroidTest`

```
:app:preBuild                                                                                            
:app:compileDebugNdk UP-TO-DATE     
:app:preDebugBuild                
:app:checkDebugManifest                
:app:prepareDebugDependencies                
:app:compileDebugAidl UP-TO-DATE     
:app:compileDebugRenderscript UP-TO-DATE      
:app:generateDebugBuildConfig UP-TO-DATE      
:app:generateDebugAssets UP-TO-DATE      
:app:mergeDebugAssets UP-TO-DATE      
:app:generateDebugResValues UP-TO-DATE      
:app:generateDebugResources UP-TO-DATE      
:app:mergeDebugResources UP-TO-DATE      
:app:processDebugManifest UP-TO-DATE      
:app:processDebugResources UP-TO-DATE      
:app:generateDebugSources UP-TO-DATE      
:core:compileJava UP-TO-DATE                                               
:core:processResources UP-TO-DATE      
:core:classes UP-TO-DATE      
:core:jar UP-TO-DATE      
:app:compileDebugJava UP-TO-DATE      
:app:unzipJacocoAgent UP-TO-DATE      
:app:instrumentDebug UP-TO-DATE                                                       
:app:preDexDebug UP-TO-DATE      
:app:dexDebug UP-TO-DATE      
:app:processDebugJavaRes UP-TO-DATE      
:app:validateDebugSigning                 
:app:packageDebug UP-TO-DATE      
:app:zipalignDebug UP-TO-DATE      
:app:assembleDebug UP-TO-DATE      
:app:compileDebugTestNdk UP-TO-DATE      
:app:preDebugTestBuild                 
:app:prepareDebugTestDependencies                 
:app:compileDebugTestAidl UP-TO-DATE      
:app:processDebugTestManifest UP-TO-DATE      
:app:compileDebugTestRenderscript UP-TO-DATE      
:app:generateDebugTestBuildConfig UP-TO-DATE      
:app:generateDebugTestAssets UP-TO-DATE      
:app:mergeDebugTestAssets UP-TO-DATE      
:app:generateDebugTestResValues UP-TO-DATE      
:app:generateDebugTestResources UP-TO-DATE      
:app:mergeDebugTestResources UP-TO-DATE      
:app:processDebugTestResources UP-TO-DATE      
:app:generateDebugTestSources UP-TO-DATE      
:app:compileDebugTestJava UP-TO-DATE      
:app:preDexDebugTest UP-TO-DATE      
:app:dexDebugTest UP-TO-DATE      
:app:processDebugTestJavaRes UP-TO-DATE      
:app:packageDebugTest UP-TO-DATE      
:app:assembleDebugTest UP-TO-DATE      
:app:forkDebug                 
Use -Dandroid.test.classes=REGEX to specify a pattern for the classes/packages to run
Use -Dfork.tablet=(true|false) to configure pools depending on their manufacturer's 'tablet' flag (ro.build.characteristics)
Use -Dfork.pool.POOL_NAME=(Serial','?)* to add devices with a given serial to a pool with given name,e.g. hdpi=01234567,abcdefgh
Use -Dfork.computed.STRATEGY=(PoolName=LowerBound','?)* to automatically create pools based on device characteristics
STRATEGY:=sw - by smallest width, e.g.phablet=0,tablet=720
STRATEGY:=api - by api, e.g. gingerbread_and_earlier=0,honeycomb_and_later=11)
Use -Dfork.eachdevice=(true|false) to create a pool per device (a.k.a. Spoon mode). This is the default behaviour.
Use -Dfork.excluded.serial=(Serial','?)* to exclude specific devices from running any tests
Use -Dfork.report.title=Title to specify a title for the generated report
Use -Dfork.report.subtitle=Subitle to specify a subtitle for the generated report
Error while Fork runner was executing
java.lang.IllegalArgumentException: Array is empty
        at org.apache.commons.lang3.BooleanUtils.and(BooleanUtils.java:929)
        at com.shazam.fork.summary.OutcomeAggregator.and(OutcomeAggregator.java:59)
        at com.shazam.fork.summary.OutcomeAggregator.access$100(OutcomeAggregator.java:27)
        at com.shazam.fork.summary.OutcomeAggregator$1.apply(OutcomeAggregator.java:42)
        at com.shazam.fork.summary.OutcomeAggregator$1.apply(OutcomeAggregator.java:36)
        at com.google.common.collect.Iterators$8.transform(Iterators.java:794)
        at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48)
        at java.util.AbstractCollection.toArray(AbstractCollection.java:195)
        at com.shazam.fork.summary.OutcomeAggregator.and(OutcomeAggregator.java:59)
        at com.shazam.fork.summary.OutcomeAggregator.aggregate(OutcomeAggregator.java:32)
        at com.shazam.fork.ForkRunner.run(ForkRunner.java:115)
        at com.shazam.fork.Fork.run(Fork.java:46)
        at com.shazam.fork.Fork$run.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
        at com.shazam.fork.gradle.ForkRunTask.runFork(ForkRunTask.groovy:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:63)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.doExecute(AnnotationProcessingTaskFactory.java:218)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:211)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:200)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:579)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:562)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:42)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.AbstractTask.executeWithoutThrowingTaskFailure(AbstractTask.java:305)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.executeTask(AbstractTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:63)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:51)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:23)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:88)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:29)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
        at org.gradle.execution.DefaultBuildExecuter.access$200(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$2.proceed(DefaultBuildExecuter.java:68)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:55)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:149)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:86)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:80)
        at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:33)
        at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:24)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:36)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:47)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:34)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:35)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:24)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.StartStopIfBuildAndStop.execute(StartStopIfBuildAndStop.java:33)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:71)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:69)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:69)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:34)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:70)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:34)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.DaemonHygieneAction.execute(DaemonHygieneAction.java:39)
        at org.gradle.launcher.daemon.server.exec.DaemonCommandExecution.proceed(DaemonCommandExecution.java:119)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:46)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:246)
        at org.gradle.internal.concurrent.DefaultExecutorFactory$StoppableExecutorImpl$1.run(DefaultExecutorFactory.java:64)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
:app:forkDebug FAILED          
              
FAILURE: Build failed with an exception.
              
* What went wrong:
Execution failed for task ':app:forkDebug'.
> Tests failed! See /Users/Blundell/MyApp/app/build/fork/debug/html/index.html
              
* Try:        
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
              
BUILD FAILED  
              
Total time: 20.536 secs
```

![screen shot 2015-02-03 at 21 42 02](https://cloud.githubusercontent.com/assets/655860/6035758/c99ca0c4-abed-11e4-8402-05e8ed525701.png)
![screen shot 2015-02-03 at 21 42 06](https://cloud.githubusercontent.com/assets/655860/6035759/c99e26d8-abed-11e4-8cfa-3265063d1d89.png)

I would have expected it to work with all default params. It saying `array is empty` makes me think it might need a argument? or looking at the source [https://github.com/shazam/fork/blob/master/src/main/java/com/shazam/fork/summary/OutcomeAggregator.java#L58](https://github.com/shazam/fork/blob/master/src/main/java/com/shazam/fork/summary/OutcomeAggregator.java#L58) it's looking for test outcomes, so has it failed to pull the results from the device?

--- a/src/main/java/com/shazam/fork/Configuration.java
+++ b/src/main/java/com/shazam/fork/Configuration.java
@@ -15,6 +15,7 @@ package com.shazam.fork;
 import com.shazam.fork.model.InstrumentationInfo;
 
 import java.io.File;
+import java.util.regex.Pattern;
 
 public class Configuration {
     private final File androidSdk;
@@ -22,17 +23,22 @@ public class Configuration {
     private final File instrumentationApk;
     private final InstrumentationInfo instrumentationInfo;
     private final File output;
+    private final Pattern testClassPattern;
+    private final Pattern testPackagePattern;
     private final int idleTimeout;
     private final int testTimeout;
     private final int testIntervalTimeout;
 
-    Configuration(File androidSdk, File applicationApk, File instrumentationApk, InstrumentationInfo instrumentationInfo,
-                  File output, int idleTimeout, int testTimeout, int testIntervalTimeout) {
+    Configuration(File androidSdk, File applicationApk, File instrumentationApk,
+                  InstrumentationInfo instrumentationInfo, File output, Pattern testClassPattern,
+                  Pattern testPackagePattern, int idleTimeout, int testTimeout, int testIntervalTimeout) {
         this.androidSdk = androidSdk;
         this.applicationApk = applicationApk;
         this.instrumentationApk = instrumentationApk;
         this.instrumentationInfo = instrumentationInfo;
         this.output = output;
+        this.testClassPattern = testClassPattern;
+        this.testPackagePattern = testPackagePattern;
         this.idleTimeout = idleTimeout;
         this.testTimeout = testTimeout;
         this.testIntervalTimeout = testIntervalTimeout;
@@ -56,6 +62,14 @@ public class Configuration {
 
     public File getOutput() {
         return output;
+    }
+
+    public Pattern getTestClassPattern() {
+        return testClassPattern;
+    }
+
+    public Pattern getTestPackagePattern() {
+        return testPackagePattern;
     }
 
     public int getIdleTimeout() {

--- a/src/main/java/com/shazam/fork/DexUtils.java
+++ b/src/main/java/com/shazam/fork/DexUtils.java
@@ -1,0 +1,18 @@
+package com.shazam.fork;
+
+public class DexUtils {
+
+    public static String getClassName(String typeDescriptor) {
+        int finalSlashIndex = getFinalSlashIndex(typeDescriptor);
+        return typeDescriptor.substring(finalSlashIndex + 1, typeDescriptor.length() - 1);
+    }
+
+    public static String getPackageName(String typeDescriptor) {
+        int finalSlashIndex = getFinalSlashIndex(typeDescriptor);
+        return typeDescriptor.substring(1, finalSlashIndex).replace('/', '.');
+    }
+
+    private static int getFinalSlashIndex(String typeDescriptor) {
+        return typeDescriptor.lastIndexOf('/');
+    }
+}

--- a/src/main/java/com/shazam/fork/ForkBuilder.java
+++ b/src/main/java/com/shazam/fork/ForkBuilder.java
@@ -9,7 +9,12 @@
  */
 package com.shazam.fork;
 
+import com.shazam.fork.model.InstrumentationInfo;
+
 import java.io.File;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
@@ -21,7 +26,9 @@ public class ForkBuilder {
     private File androidSdk;
     private File applicationApk;
     private File instrumentationApk;
-    private File output;
+    private File output = cleanFile("fork-output");
+    private Pattern testClassPattern = Pattern.compile("^((?!Abstract).)*Test$");
+    private Pattern testPackagePattern;
     private int idleTimeout = 2 * 60 * 1000; // Empirical default.
     private int testTimeout = 4 * 60 * 1000; // Empirical default.
     private int testIntervalTimeout = 30 * 1000; // Empirical default.
@@ -65,8 +72,30 @@ public class ForkBuilder {
      * @param output the output directory
      * @return this builder
      */
-    public ForkBuilder withOutputDirectory(File output) {
-        this.output = output;
+    public ForkBuilder withOutputDirectory(@Nullable File output) {
+        if (output != null) {
+            this.output = output;
+        }
+        return this;
+    }
+
+    /**
+     * Regex {@link Pattern} determining the class names to consider when finding tests to run.
+     */
+    public ForkBuilder withTestClassPattern(@Nullable Pattern testClassPattern) {
+        if (testClassPattern != null) {
+            this.testClassPattern = testClassPattern;
+        }
+        return this;
+    }
+
+    /**
+     * Regex {@link Pattern} determining the packages to consider when finding tests to run.
+     */
+    public ForkBuilder withTestPackagePattern(@Nullable Pattern testPackagePattern) {
+        if (testPackagePattern != null) {
+            this.testPackagePattern = testPackagePattern;
+        }
         return this;
     }
 
@@ -109,16 +138,26 @@ public class ForkBuilder {
         checkArgument(instrumentationApk.exists(), "Instrumentation APK file does not exist.");
         checkNotNull(output, "Output path is required.");
 
+        InstrumentationInfo instrumentationInfo = parseFromFile(instrumentationApk);
         Configuration configuration = new Configuration(
                 androidSdk,
                 applicationApk,
                 instrumentationApk,
-                parseFromFile(instrumentationApk),
+                instrumentationInfo,
                 output,
+                testClassPattern,
+                testPackagePattern,
                 idleTimeout,
                 testTimeout,
                 testIntervalTimeout);
         return new Fork(configuration);
     }
 
+    @SuppressWarnings("ReturnOfNull")
+    private static File cleanFile(String path) {
+        if (path == null) {
+            return null;
+        }
+        return new File(path);
+    }
 }

--- a/src/main/java/com/shazam/fork/ForkCli.java
+++ b/src/main/java/com/shazam/fork/ForkCli.java
@@ -18,14 +18,16 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
+import java.util.regex.Pattern;
 
 import static com.shazam.fork.ForkBuilder.aFork;
 
 public class ForkCli {
+
     private static final Logger logger = LoggerFactory.getLogger(ForkCli.class);
-    private static final String DEFAULT_OUTPUT_DIRECTORY = "fork-output";
 
     public static class CommandLineArgs {
+
         @Parameter(names = { "--sdk" }, description = "Path to Android SDK")
         public File sdk = cleanFile(System.getenv("ANDROID_HOME"));
 
@@ -38,7 +40,14 @@ public class ForkCli {
         public File testApk;
 
         @Parameter(names = { "--output" }, description = "Output path", converter = FileConverter.class)
-        public File output = cleanFile(DEFAULT_OUTPUT_DIRECTORY);
+        public File output;
+
+        @Parameter(names = { "--testClassPattern" }, description = "Regex determining class names to consider when finding tests to run", converter = PatternConverter.class)
+        public Pattern testClassPattern;
+
+        @Parameter(names = { "--testPackagePattern" }, description = "Regex determining packages to consider when finding tests to run. " +
+                "Defaults to the instrumentation package.", converter = PatternConverter.class)
+        public Pattern testPackagePattern;
 
         @Parameter(names = { "--fail-on-failure" }, description = "Non-zero exit code on failure")
         public boolean failOnFailure;
@@ -53,6 +62,14 @@ public class ForkCli {
         @Override
         public File convert(String s) {
             return cleanFile(s);
+        }
+    }
+
+    public static class PatternConverter implements IStringConverter<Pattern> {
+
+        @Override
+        public Pattern convert(String string) {
+            return Pattern.compile(string);
         }
     }
 
@@ -87,6 +104,8 @@ public class ForkCli {
                 .withInstrumentationApk(parsedArgs.testApk)
                 .withOutputDirectory(parsedArgs.output)
                 .withAndroidSdk(parsedArgs.sdk)
+                .withTestClassPattern(parsedArgs.testClassPattern)
+                .withTestPackagePattern(parsedArgs.testPackagePattern)
                 .build();
 
         if (!fork.run() && parsedArgs.failOnFailure) {

--- a/src/main/java/com/shazam/fork/TestClassScanner.java
+++ b/src/main/java/com/shazam/fork/TestClassScanner.java
@@ -12,6 +12,7 @@
  */
 package com.shazam.fork;
 
+import com.shazam.fork.model.InstrumentationInfo;
 import com.shazam.fork.model.TestClass;
 import org.jf.dexlib.AnnotationDirectoryItem;
 import org.jf.dexlib.AnnotationItem;
@@ -35,22 +36,29 @@ import static org.apache.commons.io.IOUtils.closeQuietly;
 import static org.apache.commons.io.IOUtils.copyLarge;
 
 public class TestClassScanner {
-    private static final Pattern TEST_CLASS_PATTERN = Pattern.compile("^((?!Abstract).)*Test;$");
+
     private static final String TEST = "test";
     public static final String CLASSES_PREFIX = "classes";
     public static final String DEX_EXTENSION = ".dex";
     private final File instrumentationApkFile;
     private File outputFolder;
+    private final Pattern testClassPattern;
+    private final Pattern testPackagePattern;
 
-	public TestClassScanner(File instrumentationApkFile, File outputFolder) {
-		this.instrumentationApkFile = instrumentationApkFile;
+    public TestClassScanner(File instrumentationApkFile,
+                            File outputFolder,
+                            Pattern testClassPattern,
+                            Pattern testPackagePattern) {
+        this.instrumentationApkFile = instrumentationApkFile;
         this.outputFolder = outputFolder;
-	}
+        this.testClassPattern = testClassPattern;
+        this.testPackagePattern = testPackagePattern;
+    }
 
-	public List<TestClass> scanForTestClasses() {
-		getDexClassesFromApk(instrumentationApkFile, outputFolder);
-		return getTestClassesFromDexFile(outputFolder);
-	}
+    public List<TestClass> scanForTestClasses() {
+        getDexClassesFromApk(instrumentationApkFile, outputFolder);
+        return getTestClassesFromDexFile(outputFolder);
+    }
 
     private static void getDexClassesFromApk(File apkFile, File outputFolder) {
         ZipFile zip = null;
@@ -83,7 +91,7 @@ public class TestClassScanner {
         }
     }
 
-    private static List<TestClass> getTestClassesFromDexFile(File dexFilesFolder) {
+    private List<TestClass> getTestClassesFromDexFile(File dexFilesFolder) {
         try {
             File[] dexFiles = dexFilesFolder.listFiles(new FilenameFilter() {
                 @Override
@@ -97,7 +105,9 @@ public class TestClassScanner {
                 List<ClassDefItem> items = dexFile.ClassDefsSection.getItems();
                 for (ClassDefItem classDefItem : items) {
                     final String typeDescriptor = classDefItem.getClassType().getTypeDescriptor();
-                    if (TEST_CLASS_PATTERN.matcher(typeDescriptor).matches()) {
+                    String className = DexUtils.getClassName(typeDescriptor);
+                    String packageName = DexUtils.getPackageName(typeDescriptor);
+                    if (testClassPattern.matcher(className).matches() && testPackagePattern.matcher(packageName).matches()) {
                         TestClass testClass = getTestClass(typeDescriptor);
                         testClasses.add(testClass);
                         try {

--- a/src/main/java/com/shazam/fork/injector/TestClassScannerInjector.java
+++ b/src/main/java/com/shazam/fork/injector/TestClassScannerInjector.java
@@ -16,10 +16,16 @@ import com.shazam.fork.TestClassScanner;
 
 import static com.shazam.fork.injector.ConfigurationInjector.configuredOutput;
 import static com.shazam.fork.injector.ConfigurationInjector.configuration;
+import static com.shazam.fork.injector.TestPackagePatternInjector.testPackagePattern;
 
 public class TestClassScannerInjector {
 
     public static TestClassScanner testClassScanner() {
-        return new TestClassScanner(configuration().getInstrumentationApk(), configuredOutput());
+        return new TestClassScanner(
+                configuration().getInstrumentationApk(),
+                configuredOutput(),
+                configuration().getTestClassPattern(),
+                testPackagePattern()
+        );
     }
 }

--- a/src/main/java/com/shazam/fork/injector/TestPackagePatternInjector.java
+++ b/src/main/java/com/shazam/fork/injector/TestPackagePatternInjector.java
@@ -1,0 +1,20 @@
+package com.shazam.fork.injector;
+
+import com.shazam.fork.Configuration;
+
+import java.util.regex.Pattern;
+
+import static com.shazam.fork.injector.ConfigurationInjector.configuration;
+
+public class TestPackagePatternInjector {
+
+    public static Pattern testPackagePattern() {
+        Configuration configuration = configuration();
+        Pattern testPackagePattern = configuration.getTestPackagePattern();
+        return testPackagePattern == null ? testPackagePattern(configuration) : testPackagePattern;
+    }
+
+    private static Pattern testPackagePattern(Configuration configuration) {
+        return Pattern.compile(configuration.getInstrumentationInfo().getApplicationPackage().replace(".", "\\.") + ".*");
+    }
+}

